MVC/EShopDemo/Controllers/ProductsController.cs

```diff
@@ -0,0 +1,164 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.EntityFrameworkCore;
+using EShopDemo.Models;
+
+namespace EShopDemo.Controllers
+{
+    public class ProductsController : Controller
+    {
+        private readonly EShopContext _context;
+        private readonly int _maxProductViewCount;
+
+        public ProductsController(
+            EShopContext context,
+            IConfiguration configuration)
+        {
+            _context = context;
+            _maxProductViewCount = configuration.GetValue<int>("MaxProductViewCount");
+        }
+
+        public async Task<IActionResult> Index()
+        {
+            IQueryable<Product> query = _context.Products.Include(p => p.Category).Include(p => p.Supplier);
+
+            if (_maxProductViewCount > 0)
+            {
+                query = query.Take(_maxProductViewCount);
+            }
+
+            var products = await query.ToListAsync();
+            return View(products);
+        }
+
+        public async Task<IActionResult> Details(int? id)
+        {
+            if (id == null)
+            {
+                return NotFound();
+            }
+
+            var product = await _context.Products
+                .Include(p => p.Category)
+                .Include(p => p.Supplier)
+                .FirstOrDefaultAsync(m => m.ProductId == id);
+            if (product == null)
+            {
+                return NotFound();
+            }
+
+            return View(product);
+        }
+
+        public IActionResult Create()
+        {
+            ViewData["CategoryId"] = new SelectList(_context.Categories, "CategoryId", "CategoryName");
+            ViewData["SupplierId"] = new SelectList(_context.Suppliers, "SupplierId", "CompanyName");
+            return View();
+        }
+
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<IActionResult> Create([Bind("ProductId,ProductName,SupplierId,CategoryId,QuantityPerUnit,UnitPrice,UnitsInStock,UnitsOnOrder,ReorderLevel,Discontinued")] Product product)
+        {
+            if (ModelState.IsValid)
+            {
+                _context.Add(product);
+                await _context.SaveChangesAsync();
+                return RedirectToAction(nameof(Index));
+            }
+            ViewData["CategoryId"] = new SelectList(_context.Categories, "CategoryId", "CategoryName", product.CategoryId);
```

Comment:
Well, `ViewData` is not really a good option since you have to perform casts/checks for key existence, instead, strongly advise for the future to use strongly-types views